### ISSUE TITLE
Aiccu Support

### ISSUE THREAD

**Issue by sjorge** [opnsense/core#57]
It would be nice to bundle aiccu to also over ipv6 tunnels that use AYIYA instead of heartbeat.

**Comment by michbu** [comment]
+1

**Comment by fichtner** [comment]
Will add it as an optional port, but somebody should help with progressing the GUI integration (if needed).

**Comment by fichtner** [comment]
I've pushed the client to the package mirror (as soon as 15.1.10 is released today). To install:

    # pkg install sixxs-aiccu

The client itself is from 2007, don't expect any wonders. I've also disabled GNUTLS support due to heavy GNU dependencies and I don't want another SSL library in the packages. It should be rewritten to use OpenSSL/LibreSSL by default (if there is enough demand for this I'll do it). Closing this for now. Cheers! :)

**Comment by mibuthu** [comment]
Thanks for inclusion, it is working fine.

The most important missing part is that the interface isn't available in the GUI, so no firewall rules can be set for it. But I found some good hints in [this thread](https://forum.pfsense.org/index.php?topic=62136.0) in the pfsense forum.

I want to do the full GUI integration. I have already tested a bit, it should relatively easy.

But before I goto the details I have some open questions I would like to discuss:
* The interface which will be added is called "tun0" independent of the "ipv6_interface" option value in the config file. There is also a hint in the config file that on BSD the ipv6_interface should be set to gifX for proto-41 tunnels or tunX for AYIYA tunnels. In the link above there is already given a fix for the GUI to include tun interfaces in the web interface. This is working fine but it is also proposed to rename the interface to a more unique name like "aiccu". I would prefer this also, aiccu should respects the given interface name in the config file and I then would use "aiccu" as the name. Would this be a problem? If yes, can the interface name "tun0" can cause problems with other interfaces that are named tunX or is it o.k. to add all tun interfaces to the web UI?
* I would add an additional tab in the interfaces assign screen called "Aiccu". In all other tabs there is an overview of all available interfaces. For aiccu it is only possible to set 1 interface, so the question is if it make sense to have an overview page. I can only allow to create 1 single interface. The alternative would be to remove the overview page at all and only show the options with an enable/disable checkbox. What do you prefer?
* Do you have any other suggestions which I should respect?

Of course it would be great if you can do the SSL parts described above.

**Comment by fichtner** [comment]
@mibuthu thanks for the confirmation and your help :)

(a) I don't thin renaming is a particularly good idea as it creates inconsistencies and ambiguities in the long run. The actual type and configuration of an interface should not be put into the name just for the sake of picking it up in the GUI. It creates more confusion than it solves, especially when trying to debug a tun interface that is not there. We can certainly add tun devices, but need to make sure OpenVPN devices which also use tun are not getting in the way (they are currently renamed too but I think this should go away as well as renaming vlans).
(b) Maybe we can go a different route and not do the interface setup inside the interface but simply add a "service" for aiccu if there can only be one interface anyway and hide "interface" complexity and idiosyncrasies completely... or call it Tunnel and shuffle the VPN functionality into this...
(c) Only one suggestion: please don't add static PHP pages and instead use the MVC to implement this feature as we are trying to migrate everything. It's a long process, but not building static pages is a good way to eventually switch over.

So the SSL part is essential? Just checking.

I'll reopen this ticket then, would love to discuss further. :)


Cheers,
Franco

**Comment by mibuthu** [comment]
Thanks for the info.
(a) Noticed, so I will do a first commit where I add tun interfaces to the network interfaces handled by the gui.
(b) Maybe your right, so I would propose to add a new service menu point called "Aiccu" and probably also a new status page called "Aiccu" which shows the status of the tunnel (something like you get with sixxs-aiccu autotest). Probably this is the best solution because it is a service which creates a tun or gif interface. Also your alternative proposal would also be suitable (if I understand you correctly renaming the "VPN" menu point to "Tunnels" and add Aiccu there) but then in my opinion more parts from interfaces should be moved to this section (e.g. gre and gif interface pages which are also tunnels).
(c) Ok, I can do that, are there already some similar pages converted/available in MVC concept which I can use as a boilerplate?
(d) I'm not a pro in this topic, but in my opinion the SSL part is important, because e.g. also username and password are transfers.

Can you please comment again, afterwards I will start coding.

Thanks in advance.

**Comment by fichtner** [comment]
Sorry, this slipped through due to a few days of vacation.

(a) nice :)
(b) agreed to the initial approach, the "tunnel" rework will come much later as there is other things to do first than shuffle menu items around for mostly ease of use.
(c) The proxy is a good example for the MVC usage, although a bit complex. We'll add a better sample/tutorial with 15.7 so that others will have an easier time in the future. https://github.com/opnsense/core/tree/master/src/opnsense/mvc/app/controllers/OPNsense/Proxy
(d) ack, I will evaluate how much work that'll be

Cheers,
Franco

**Comment by sjorge** [comment]
For now it is only command line right?

**Comment by fichtner** [comment]
Yes.

**Comment by 8191** [comment]
@mibuthu Are you working on the GUI? If not, do you mind if I grab this task?

**Comment by AdSchellevis** [comment]
@8191 if you need help using the MVC framework, just let me know.

**Comment by 8191** [comment]
@AdSchellevis I'll keep that in mind, thanks. But thanks to your HelloWorld app, the first step is not that big anymore... ;)

**Comment by mibuthu** [comment]
@8191 Not really. I didn't find time to work on it, yet. So feel free to grab it.
If you need help please tell me, probably I can do a part of it.

**Comment by 8191** [comment]
@AdSchellevis Shall this be implemented in the namespace `OPNsense` or better within something like `SixXS`? Or: will this potentially be part of core or more likely as a future plugin?

**Comment by AdSchellevis** [comment]
@8191 I would like to suggest to use OPNsense for plugins developed within the community/project, so third party vendors can pick their own (company) name for custom modules.
I rather won't make this part of core, because we do have a working plugin system in place now. In the future some portions in the core of the system might move to plugins as well.
If you need help with the plugin (or mvc) at some point, I can certainly help.

**Comment by 8191** [comment]
I'm running into some general issues: aiccu uses the same syslog facility as dhcpd (`user7`), so the logs get mixed up. Any idea how to resolve this? Unfortunately this is hard-coded into aiccu.

**Comment by AdSchellevis** [comment]
We should be able to filter the aiccu messages with syslog into it's own file, just like some other services do. At the moment there is no pluggable system in place to support it, but if we add the handle in core we can move it to something better later.

The output of syslog is generated to /var/etc/syslog.conf using https://github.com/opnsense/core/blob/master/src/etc/inc/system.inc#L813-L839 
(I refactored it a bit lately so it might look a bit different, adding another entry in syslogconfs will probably do the trick)

**Comment by 8191** [comment]
@AdSchellevis okay, so I'll create a PR to add this to the core.
Is there already a module in place (e.g. configd module) to allow retrieving logs via the MVC system? If not, what is your recommendation to display logs of MVC apps? Creating a configd action for that, or using the old static PHP method for now?

**Comment by AdSchellevis** [comment]
@8191 No, there isn't one yet, but I don't want to pull in legacy code into the new code base (it will make things very hard in the end to move away from the legacy parts). Let me create a simple configd call for you this week with some simple (optional) parameters.

**Comment by fichtner** [comment]
Flat log files have templates now, it's ok to add them until the template is translated to MVC. Proxy does it this way too. Suricata does something completely different, but it's more comparable to firewall log output anyway, not the general service log file pages. But that's digressing.

TL;DR: I'm ok with a old-style log file based on the template.

**Comment by 8191** [comment]
Thanks for clarification...

One more question (does not really fit in here): why are all log files prefixed with `%`? This selector is not described in `syslog.conf(5)`. What's it purpose, please?

**Comment by fichtner** [comment]
This is the clog support indicator (circular buffer). The documentation isn't very good, I had to check the source code... https://github.com/opnsense/ports/blob/master/opnsense/syslogd/files/extra-clog-support.diff

(clog(8) support comes from DragonFly BSD)

**Comment by 8191** [comment]
Ah, that explains why `tail -f` does not really work as intended... :wink:

**Comment by fichtner** [comment]
clog -f then :)

**Comment by fichtner** [comment]
FWIW, we want to go back to clog-less logfiles, but that is something for 16.7 to achieve. Log rotation needs to be improved and the MVC framework should be in place.

**Comment by 8191** [comment]
I just recognized that `syslogd` does not create non-existing log files (since they need to be initialized by `clog -i`. Is there an automated process to create log files in OPNsense?

**Comment by fichtner** [comment]
Please don't use clog. Log files are created on boot if necessary, in suricata/proxy integrtion we use this approach to bootstrap files,commands, directories: https://github.com/opnsense/core/blob/master/src/opnsense/scripts/proxy/setup.sh

**Comment by 8191** [comment]
I'm having some troubles in assigning the aiccu tun/gif interface to OPNsense, so that one can define firewall rules, but OPNsense does not try to set IP addresses on the interface so that aiccu still works.

Currently it's possible to assign the "tun0" as an interface if it was generated by aiccu (due to #284), but then aiccu cannot correctly establish the tunnel anymore, since a script seems to interfere with the IP address assignment... How are IP addresses actually being set through OPNsense?

I could implement it in a similar way as OpenVPN and IPSec firewall rules are defined, but since the firewall is still an "old" PHP script, there is no possibility to dynamically add interface and we would need to have parts of the aiccu interface logic within the `www/firewall_rules.php`.

Any other ideas?

**Comment by AdSchellevis** [comment]
Is there a simple way I could test some of your code already? We should look at some pluggable structure to define interfaces, but first the old code needs to be cleaned-up for this area.
My plan is to start cleaning the interfaces frontend code later this week, when the code quality is better we may integrate something new there to facilitate these scenarios.
Hard-coding the different connectivity types is probably not the best solution in the world....

**Comment by 8191** [comment]
@AdSchellevis the current state is checked-in in [8191/opnsense-core@aiccu](https://github.com/8191/opnsense-core/tree/aiccu)

**Comment by AdSchellevis** [comment]
@8191 I haven't found time to test your code yet, I do want to cleanup the interface section first to retrieve some clarity on the subject.
One question, how do you assign a network address normally for aiccu? if ifconfig does the trick, we should be able to hook in something to register new interfaces for the legacy code to use. if the daemon handles the assignment, it probably needs another approach.

**Comment by sjorge** [comment]
IIRC you give it a gif or tun (depending on how you configure AICCU) and the daemon manages the addresses

**Comment by 8191** [comment]
Exactly, you define a tun/gif (e.g. `tun9`) and aiccu creates the interface and manages the IP. But there also exists a config option to let the user take care of the IP configuration (e.g. using ifconfig). 

I currently have troubles defining the interface. Shall the interface be hard-coded  (e.g. always `tun0`)? Or shall it be determined using a specific OPNsense mechanism? Or are there even use cases where the user shall be able to define the interface?

**Comment by 8191** [comment]
@fichtner You can assign that issue to me, please. It's [already in the queue](https://github.com/8191/opnsense-core/commits/aiccu), but I'm not sure how to solve the FW rule assignment yet.

**Comment by fichtner** [comment]
if anybody wants to pick this up let me know
